### DISCUSSION TITLE
Add new device class: frequency

### DISCUSSION THREAD

**Discussion by andriej** [home-assistant/architecture#600] [Entity Models] [answered]
As grid (and electricity) is now on top of everyones head - I've also integrated my electricity meter to HA.
And was kinda surprised, that Frequency sensor from that meter (for EU it should be 50Hz) doesn't have it's own device class.
Hertz (and MegaHertz, GigaHertz etc.) 

It's not only electricity meters that would benefit from this device class, as it could be set for every sensor that shows how often per second something is happening.

**Comment by MartinHjelmare** [comment]
To better take a decision on this new device class we'd like to know how many integrations could benefit from this addition. Can you research this?

**Reply by glcos** [reply]
The PR I'm proposing requires this device class
https://github.com/home-assistant/core/pull/59417

**Reply by frenck** [reply]
@glcos Yes, but that doesn't answer the question you are responding to...

**Comment by epenet** [comment]
`grep -r "FREQUENCY" **/sensor.py` shows these would likely benefit from the new device class (they have sensors which use the frequency units):
- apcupsd
- cpuspeed
- daikin
- homematic
- huawei_lte
- iotawatt
- mysensors
- solaredge
- system_bridge
- tasmota
- zha (in [home-assistant/core#59417](https://github.com/home-assistant/core/pull/59417))

These two I don't think match the criteria
- lacrosse
- powerwall

**Reply by MartinHjelmare** [reply]
With 11 integrations that can benefit, I'm in favor of adding this device class. :+1:

**Reply by Hedda** [reply]
[Z-Wave JS integration](https://www.home-assistant.io/integrations/zwave_js/) could also benefit in the future as frequency is feature of sensors in many existing physical Z-Wave electricity meter devices (like clamp power meters) that you can buy, (e.g. maybe [Aeotec Home Energy Meter](https://aeotec.com/z-wave-home-energy-measure/index.html) and [Qubino 3-Phase Smart Meter](https://qubino.com/products/3-phase-smart-meter/)).

Probably integrations for [other hubs/gateways/bridges](https://www.home-assistant.io/integrations/#hub) when they have a Z-Wave radio, such as maybe the [SmartThings integration](https://www.home-assistant.io/integrations/smartthings/)(?).

At least there exists a scale in the Z-Wave specifications for Frequency in Hz or KHz. See zwave-js upstream question and answer [here.](https://github.com/zwave-js/node-zwave-js/discussions/3695)

**Comment by Hedda** [comment, marked as answer]
FYI, epenet has now submitted pull requests for adding "frequency" (Hz, MHz, or GHz) as a new device class in Home Assistant core and UI:

https://github.com/home-assistant/core/pull/59700

https://github.com/home-assistant/frontend/pull/10621

https://github.com/home-assistant/developers.home-assistant/pull/1137

https://github.com/home-assistant/home-assistant.io/pull/20319

**Comment by Hedda** [comment]
As mentioned in the HA core PR; please consider also adding "kHz" (kilohertz) as well? 

Today only "Hz" and "KHz" exists as scales for Frequency in the Z-Wave specifications, see Z-Wave JS -> zwave-js/node-zwave-js#3695

**Reply by Hedda** [reply]
hmm @epenet pointed out that `kHz` is not (yet?) part of the standard units in const.py:

https://github.com/home-assistant/core/blob/5370dd81226853c7c8e6b2c5b7310c38a3e0948a/homeassistant/const.py#L472-L475

```
 # Frequency units 
 FREQUENCY_HERTZ: Final = "Hz" 
 FREQUENCY_MEGAHERTZ: Final = "MHz" 
 FREQUENCY_GIGAHERTZ: Final = "GHz"
```

Can "kHz" that be just added as an additional standard unit in const.py for PR https://github.com/home-assistant/core/pull/59700 or is a new architecture discussion and decision needed specifically for that? https://github.com/home-assistant/architecture/discussions/


```
 FREQUENCY_KILOHERTZ: Final = "kHz"
```

Unit for `kHz` would be needed to stay compliant with Z-Wave specification without having to do the conversion in the integration.

**Reply by epenet** [reply]
As specified by @frenck, that would introduce dead code.
It is better to add it when the first component (zwave or another) is ready to make use of the new device class.

**Reply by Hedda** [reply]
> As specified by frenck, that would introduce dead code. It is better to add it when the first component (zwave or another) is ready to make use of the new device class.

OK so it was a what comes first issue, chicken or the egg. Nice! See that you emended the comment about it in the test sensor. Thanks!

**Reply by Hedda** [reply]
Ah, cool, raman325 was quick to take advantage of that and started working on unit mapping for zwave_js meters and sensors:

https://github.com/home-assistant/core/pull/59659

What came first, chicken or the egg, egg or the chicken ;)
